asl_xdsl/dialects/asl.py

```diff
@@ -1526,6 +1555,149 @@ def __init__(
+    """
+    The type of a reference to an object
+    """
+
+    name = "asl.ref"
+    type: ParameterDef[Attribute]
+
+    def print_parameters(self, printer: Printer) -> None:
+        # We need this to pretty print a tuple and its members if
+        # this is referencing one, otherwise just let the type
+        # handle its own printing
+        printer.print("<")
+        if isinstance(self.type, TupleType):
+            printer.print("tuple<")
+            for idx, t in enumerate(self.type.types.data):
+                if idx > 0:
+                    printer.print(", ")
+                printer.print(t)
+            printer.print(">")
+        else:
+            printer.print(self.type)
+        printer.print(">")
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
+        # This is complicated by the fact we need to parse tuple
+        # here also as the buildin dialect does not support this
+        # yet
+        parser.parse_characters("<")
+        has_tuple = parser.parse_optional_keyword("tuple")
+        if has_tuple is None:
+            param_type = parser.parse_type()
+            parser.parse_characters(">")
+            return [param_type]
+        else:
+            # If its a tuple then there are any number of types
+            def parse_types():
+                return parser.parse_type()
+
+            param_types = parser.parse_comma_separated_list(
+                parser.Delimiter.ANGLE, parse_types
+            )
+            parser.parse_characters(">")
+            return [TupleType(param_types)]
+
+    @classmethod
+    def constr(
+        cls,
+        type: IRDLAttrConstraint | None = None,
+    ) -> GenericAttrConstraint[ReferenceType]:
+        if type is None:
+            return BaseAttr[ReferenceType](ReferenceType)
+        return ParamAttrConstraint[ReferenceType](
+            ReferenceType,
+            (type,),
+        )
+
+
+@irdl_op_definition
+class GlobalOp(IRDLOperation):
+    name = "asl.global"
+
+    assembly_format = "$sym_name `:` $global_type attr-dict"
+
+    global_type = prop_def(Attribute)
+    sym_name = prop_def(StringAttr)
+
+    traits = traits_def(SymbolOpInterface())
+
+    def __init__(
+        self,
+        global_type: Attribute,
+        sym_name: str | StringAttr,
+    ):
+        if isinstance(sym_name, str):
+            sym_name = StringAttr(sym_name)
+
+        props: dict[str, Attribute] = {
+            "global_type": global_type,
+            "sym_name": sym_name,
+        }
+
+        super().__init__(properties=props)
+
+
+@irdl_op_definition
+class AddressOfOp(IRDLOperation):
+    """
+    Convert a global reference to an SSA-value to be
+    used in other operations.
+
+    %p = asl.address_of(@symbol) : !asl.ref<i1>
+    """
+
+    name = "asl.address_of"
+    assembly_format = "`(` $symbol `)` `:` type($res) attr-dict"
+    symbol = prop_def(SymbolRefAttr)
+    res = result_def()
```

Comment:
It really doesn't matter in which order the things that aren't `.*_def` in the definition, but I got quite used to the following, using this in this file may help future readability:

```suggestion
    name = "asl.address_of"
    
    symbol = prop_def(SymbolRefAttr)
    res = result_def()
    
    assembly_format = "`(` $symbol `)` `:` type($res) attr-dict"
```

So name, followed by newline, then all the defs, then newline, then assembly format, grouping them by kind. This is definitely a nit so only if you agree :)